### PR TITLE
Use ranges to keep track of merge join queries

### LINKED ISSUE #2730: Allow merge join queries to skip ahead
At the moment, the skip ahead feature for merge joins is not working. If there are no suitable entries in a given chunk for one of the ranges, the scan will stop. For example:

Range 1:
```
value1/key00001
...[10,000 keys later]
value1/key10002
```

Range 2:
```
value2/key10002
```

After fetching the first chunk from range 1, the query does not skip ahead like it should (to the first key in range 2) since there are no relevant entries from the range.

### PR DESCRIPTION
**Pull request by cdonati**
This allows the matcher to skip ahead with only the ranges that need to be adjusted. It also reduces the index entry fetch size from 10,000 to 1000.

Resolves #2730

### CODE REVIEW

File: AppDB/appscale/datastore/range_iterator.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    start_key = KEY_DELIMITER.join([self.prefix, str(encode_index_pb(path))])
+    end_key = ''.join([start_key, TERMINATING_STRING])
+    if start_key < self._range[0] or end_key > self._range[-1]:
+      raise BadRequest('Restriction must be within range')
+
+    if self._cursor.key > end_key:
+      raise BadRequest('Cursor already exceeds new range')
+
+    self._range = (start_key, end_key)
+    self._cursor.key = max(start_key, self._cursor.key)
```
Comment (sjones4) [inline AppDB/appscale/datastore/range_iterator.py]: isn't self._cursor immutable? (namedtuple)

File: AppDB/appscale/datastore/range_iterator.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+class RangeIterator(object):
+  """ Iterates through a range of index entries.
+
+  This was designed for merge join queries. The range can only be narrowed.
+  """
+  CHUNK_SIZE = 1000
+
+  def __init__(self, db, project_id, namespace, kind, prop_name, value):
+    """ Creates a new IndexReader.
```
Comment (whoarethebritons) [inline AppDB/appscale/datastore/range_iterator.py]: RangeIterator

Comment (whoarethebritons) [inline AppDB/appscale/datastore/range_iterator.py]: RangeIterator

File: AppDB/appscale/datastore/range_iterator.py (hunk 1 above)
Comment (cdonati) [reply AppDB/appscale/datastore/range_iterator.py]: Whoops. You're right. It also looks like we weren't testing that code path. https://github.com/AppScale/hawkeye/pull/140 should cover that case.

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/5152/

**Comment by cdonati** [comment]
This branch with https://github.com/AppScale/hawkeye/pull/140: https://ocd.appscale.com:8080/job/Hawkeye/14930/

**Comment by scragraham** [comment]
demo looks good

**Pull request merged**